crates/planner/src/logical/bind.rs

```diff
@@ -0,0 +1,414 @@
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        Type::Alg(AlgebraicType::U64) => v
+            .parse::<u64>()
+            .map(AlgebraicValue::U64)
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        Type::Alg(AlgebraicType::F32) => v
+            .parse::<f32>()
+            .map(|v| AlgebraicValue::F32(v.into()))
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        Type::Alg(AlgebraicType::F64) => v
+            .parse::<f64>()
+            .map(|v| AlgebraicValue::F64(v.into()))
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        Type::Alg(AlgebraicType::I128) => v
+            .parse::<i128>()
+            .map(|v| AlgebraicValue::I128(v.into()))
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        Type::Alg(AlgebraicType::U128) => v
+            .parse::<u128>()
+            .map(|v| AlgebraicValue::U128(v.into()))
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        Type::Alg(t) if t.is_bytes() => from_hex_pad::<Vec<u8>, _>(&v)
+            .map(|v| AlgebraicValue::Bytes(v.into_boxed_slice()))
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        Type::Alg(t) if t.is_identity() => Identity::from_hex(&v)
+            .map(AlgebraicValue::from)
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        Type::Alg(t) if t.is_address() => Address::from_hex(&v)
+            .map(AlgebraicValue::from)
+            .map(|v| Expr::Lit(v, ty.clone()))
+            .map_err(|_| constraint_err(&v, ty)),
+        _ => Err(constraint_err(&v, ty)),
+    }
+}
+
+/// Returns a type constraint violation for an unexpected type
+fn unexpected_type(expected: &Type, actual: &Type) -> TypingError {
+    ConstraintViolation::eq(expected, actual).into()
+}
+
+/// Returns an error if the input type is not a table type [Type::Var]
+fn expect_table_type(expr: RelExpr) -> TypingResult<RelExpr> {
+    match expr.ty() {
+        Type::Var(_) => Ok(expr),
+        _ => Err(Unsupported::SubReturnType.into()),
+    }
+}
+
+/// Assert that this type is compatible with this operator
+fn expect_op_type(op: BinOp, expr: Expr) -> TypingResult<Expr> {
+    match (op, expr.ty()) {
+        // Logic operators take booleans
+        (BinOp::And | BinOp::Or, Type::Alg(AlgebraicType::Bool)) => Ok(expr),
+        // Comparison operators take integers or floats
+        (BinOp::Lt | BinOp::Gt | BinOp::Lte | BinOp::Gte, Type::Alg(t)) if t.is_integer() || t.is_float() => Ok(expr),
+        // Equality supports numerics, strings, and bytes
+        (BinOp::Eq | BinOp::Ne, Type::Alg(t))
+            if t.is_bool()
+                || t.is_integer()
```

Comment:
I think we should add a trait similar to `eq, ord` so we instead do `if t.is_comparable`